File: sdk/runanywhere-flutter/packages/runanywhere/lib/public/capabilities/runanywhere_plugin_loader.dart
// SPDX-License-Identifier: Apache-2.0
//
// runanywhere_plugin_loader.dart — Plugin Loader capability surface
// (canonical §12 namespace). Mirrors Swift `RunAnywhere.PluginLoader`
// and the RN/Web `RunAnywhere.pluginLoader.*` namespace.
//
// §15 type-discipline: all `dart:ffi` work lives in
// `lib/native/dart_bridge_plugin_loader.dart`; this capability calls
// into that bridge.

import 'package:runanywhere/foundation/error_types/sdk_exception.dart';
import 'package:runanywhere/internal/sdk_state.dart';
import 'package:runanywhere/native/dart_bridge_plugin_loader.dart';
import 'package:runanywhere/native/types/basic_types.dart' show RacResultCode;

/// Plugin Loader capability surface (canonical §12 namespace).
///
/// Access via `RunAnywhereSDK.instance.pluginLoader`.
class RunAnywherePluginLoaderCapability {
  RunAnywherePluginLoaderCapability._();
  static final RunAnywherePluginLoaderCapability _instance =
      RunAnywherePluginLoaderCapability._();
  static RunAnywherePluginLoaderCapability get shared => _instance;

  /// Compile-time plugin API version this build was built against.
  int get apiVersion => DartBridgePluginLoader.apiVersion();

  /// Total number of plugins currently registered.
  int get registeredCount => DartBridgePluginLoader.registeredCount();

  /// Snapshot of currently-registered plugin names.
  List<String> registeredNames() => DartBridgePluginLoader.registeredNames();

  /// Load a shared library at [path] and register the
  /// `rac_plugin_entry_<stem>` it exposes with the in-process plugin
  /// registry.
  Future<PluginInfo> load(String path) async {
    if (!SdkState.shared.isInitialized) {
      throw SDKException.notInitialized();
    }
    final result = DartBridgePluginLoader.loadPlugin(path);
    if (!result.success) {
      _throwForCode('rac_registry_load_plugin', result.resultCode);
    }
    return PluginInfo(name: result.name, path: path);
  }

  /// Unregister a previously-loaded plugin and `dlclose` its handle.
  Future<void> unload(String name) async {
    if (!SdkState.shared.isInitialized) {
      throw SDKException.notInitialized();
    }
    final rc = DartBridgePluginLoader.unloadPlugin(name);
    if (rc != 0) {
      _throwForCode('rac_registry_unload_plugin', rc);
    }
  }

  static Never _throwForCode(String op, int code) {
    if (code == RacResultCode.errorFeatureNotAvailable ||
        code == RacResultCode.errorNotImplemented) {
      throw SDKException.featureNotAvailable('PluginLoader: $op');
    }
    throw SDKException.modelLoadFailed(
      op,
      RacResultCode.getMessage(code),
    );
  }
}

/// Plugin descriptor returned from `pluginLoader.load(...)`.
class PluginInfo {
  final String name;
  final String path;
  const PluginInfo({required this.name, required this.path});
}

File: sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/public/extensions/RunAnywherePluginLoader.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Runtime plugin loader — namespaced as `RunAnywhere.pluginLoader.*`
 * per CANONICAL_API.md §12.
 *
 * Round 2 KOTLIN: Replaced flat API (loadPlugin, unloadPlugin, etc.)
 * with the canonical `RunAnywhere.pluginLoader: PluginLoader` namespace,
 * matching Swift, Flutter, RN and Web SDK surfaces.
 */

package com.runanywhere.sdk.public.extensions

import com.runanywhere.sdk.foundation.errors.SDKException
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge
import com.runanywhere.sdk.public.RunAnywhere
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

// PluginInfo — lightweight descriptor for a loaded plugin.

/**
 * Descriptor for a plugin loaded at runtime.
 *
 * @param name Plugin name / library stem (without `lib` prefix or extension, e.g. "runanywhere_acmevoice")
 * @param path Absolute path to the shared library on disk, if known
 */
data class PluginInfo(
    val name: String,
    val path: String? = null,
)

// PluginLoader — namespaced capability class

// RunAnywhere.pluginLoader accessor

class PluginLoader {
    val apiVersion: UInt
        get() = RunAnywhereBridge.racRegistryGetPluginApiVersion().toUInt()

    val registeredCount: Int
        get() = RunAnywhereBridge.racRegistryGetPluginCount()

    suspend fun load(path: String): PluginInfo =
        withContext(Dispatchers.IO) {
            // The C loader stores the plugin under vt->metadata.name (e.g. "llamacpp"
            // for the in-tree librunanywhere_llamacpp.so), which is what
            // rac_registry_unload_plugin expects. Deriving the name from the file
            // stem locally is fragile because the C++ loader also strips an optional
            // `runanywhere_` infix, and third-party plugins may set metadata.name to
            // anything they choose. Snapshot registry keys before/after the load and
            // use the diff to identify the newly registered name (mirrors Swift's
            // `RunAnywhere+PluginLoader.swift:72-89`).
            val before = registeredNames().toSet()

            val rc = RunAnywhereBridge.racRegistryLoadPlugin(path)
            if (rc != RunAnywhereBridge.RAC_SUCCESS) {
                throw SDKException.operation("rac_registry_load_plugin failed with rc=$rc")
            }

            val after = registeredNames()
            val registeredName = after.firstOrNull { it !in before } ?: deriveStemName(path)
            PluginInfo(name = registeredName, path = path)
        }

    /**
     * Mirror the C loader's stem derivation (strip directory, the optional `lib`
     * prefix, the file extension, and the optional `runanywhere_` infix). Used as
     * a last-resort fallback when the registry-diff is empty — e.g. when an
     * already-loaded plugin's entry is idempotently re-registered and
     * [registeredNames] returns the same set.
     */
    private fun deriveStemName(path: String): String {
        var stem = path.substringAfterLast('/').substringBeforeLast('.')
        if (stem.startsWith("lib")) {
            stem = stem.removePrefix("lib")
        }
        if (stem.startsWith("runanywhere_")) {
            stem = stem.removePrefix("runanywhere_")
        }
        return stem
    }

    suspend fun unload(name: String) =
        withContext(Dispatchers.IO) {
            val rc = RunAnywhereBridge.racRegistryUnloadPlugin(name)
            if (rc != RunAnywhereBridge.RAC_SUCCESS) {
                throw SDKException.operation("rac_registry_unload_plugin failed with rc=$rc")
            }
        }

    suspend fun registeredNames(): List<String> =
        withContext(Dispatchers.IO) {
            RunAnywhereBridge.racRegistryGetRegisteredNames()?.toList() ?: emptyList()
        }

    suspend fun listLoaded(): List<PluginInfo> =
        registeredNames().map { PluginInfo(name = it, path = "") }
}

// Singleton instance — one per SDK singleton.
private val pluginLoaderInstance = PluginLoader()

val RunAnywhere.pluginLoader: PluginLoader
    get() = pluginLoaderInstance
